packages/vite/src/node/preview.ts

```diff
@@ -97,19 +98,41 @@ export async function preview(
 
   app.use(compression())
 
+  const isSPA = config.isSPA ?? true
+
+  if (isSPA) {
+    // We need to apply the plugins' server post hooks before `sirv()`. (Because
+    // `sirv(_, { single: true })` catches all routes.)
+    postHooks.forEach((fn) => fn && fn())
+  }
+
   // static assets
   const distDir = path.resolve(config.root, config.build.outDir)
   app.use(
     config.base,
     sirv(distDir, {
       etag: true,
       dev: true,
-      single: true
+      single: isSPA
     })
   )
 
-  // apply post server hooks from plugins
-  postHooks.forEach((fn) => fn && fn())
+  if (!isSPA) {
+    // We apply the plugins' server post hooks after `sirv()` so that `sirv()`
+    // can serve pre-rendered pages before any SSR middleware.
+    postHooks.forEach((fn) => fn && fn())
```

Comment:
Maybe I'm missing some context but couldn't these hooks be always applied before sirv? And if every hook ignores pre-rendered pages then they will be served by sirv? Looks more flexible so even for pre-rendered pages the hooks could do some special handling. It means checking if the file exists in disk, no?
I like the idea of static assets / 404 handling being the last fallback after the postHooks in both modes.

Reply:
The thing is that SSR middlewares actually do not ignore pre-rendered pages: a pre-rendered page `/about` has been rendered at build-time to `dist/client/about.html` but if the user were to `$ rm ./dist/client/about.html` then the SSR middleware will pick up and render the page at run-time.

That's at least how vite-plugin-ssr works and I expect other frameworks to do this as well. (Because it's a canonical way to support partial pre-rendering.)

Thus, `sirv()` needs to run before the SSR middleware, otherwise the SSR middleware will wrongfully render the pre-rendered pages and `sirv()` never runs. This breaks `$ vite preview` in the sense that it's not a truthful representation of production anymore.

Reply:
Again, I may be missing some context, so thanks for the explanations. Why SSR middlewares can't check if the file exists before deciding what to do? Doesn't something similar would happen in the server where you have a chance to intercept a request and do whatever you want (even if it is pre-rendered, you may want to ignore it depending on some other variables?). @benmccann @antfu maybe you could help here, I'm just looking to avoid cornering us into fighting against sirv later. We aren't going to be able to add a pre/post middleware strategy here, so good to get this one right.

Reply:
Previously, the post hooks were always being run after sirv. You're talking about the SSR case, but it seems the case that has changed is the SPA case and the hooks are now being run before sirv in that case. Why does the SPA case need to change? I see the comment explaining it needs to happen first because `single` will catch everything, but then I don't understand why it's happening after in `main` today. Is `main` broken? Or am I reading the code incorrectly? (very possible as it's time for me to go to bed...)

Reply:
Turns out my Vitepress PR (https://github.com/vuejs/vitepress/pull/625) doesn't work for URLs ending with `.html`. I'm not sure how to fix Vitepress so I decided to revert all changes related to SPA mode.

The SPA mode of this PR now behaves exactly as `main`.

As you can see the PR now changes almost nothing: it's mostly about setting `single` to `false` for SSR (this change was already approved & merged before but got reverted because it needs to be applied only in SSR mode). The rest of the changes are actually optional, but I guess they make sense (e.g. Ben's point that some SPA middlewares are superfluous for SSR).

So I guess it's ready for merge, finally :-).

As for SSR, let me elaborate.

In general, apps are usually served like this:
 - First the URL is checked against the Static Host.
 - If the Static Host doesn't return a response, then forward the HTTP request to the Node.js server.

That's how reverse proxies e.g. NGNIX are set up. It's also how the routing part of Cloudflare workers are written. This seems standard.

Translated to this PR:
 - First `sirv()`
 - If `sirv()` doesn't deliver an HTTP response, then the SSR middleware.

In other words, the `sirv()` middleware comes first, then post hooks middlewares.

Reply:
I'm confident all concerns are addressed; I'm marking this as resolved.